<style>
h1 {
    border-bottom: 2px solid;
}
h2 {
    border-bottom: 1px solid;
}
/* hr {
    border-top: 2px solid;
    border-bottom: none;
} */
.student-info {
    text-align: right;
    border-bottom: none;
}
hr.double-div {
    border-top: 2px solid;
    border-bottom: 2px solid;
    border-left: none;
    border-right: none;
    height: 10pt;
    /* border-radius: 1px; */
}
blockquote {
    border-left: 2px solid;
    /* background-color: #00000488; */
    /* filter: brightness(80%); */
    margin-left: 2em;
    margin-right: 2em;
    padding: 1px 2px 1px 1em;
}
pre {
    border: 1px solid #88888888;
    margin-left: 1em;
    margin-right: 1em;
    padding: 0.5em;
    /*padding: 8pt 1em 8pt 1em;*/
}
blockquote > pre {
    margin-left: 0.5em;
}
img {
    display: block;
    margin-left: auto;
    margin-right: auto;
}
</style>

<h2 class="student-info">
    Andrew Osterhout <br/>
    (u1317172)
</h2>

# Asg-10 focuses on questions graded by the mentioned TAs

 

In [1]:
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
import sys

# -- Detect if in Own Install or in Colab
try:
    import google.colab
    OWN_INSTALL = False
except:
    OWN_INSTALL = True
    
if OWN_INSTALL:
    print("Own install")
    sys.path[0:0] = ['../../../../..',  '../../../../../3rdparty',
                     '../../../..',  '../../../../3rdparty',
                     '../../..',     '../../../3rdparty',
                     '../..',        '../../3rdparty',
                     '..',           '../3rdparty' ]
    
else: # In colab
    ! if [ ! -d Jove ]; then git clone https://github.com/ganeshutah/Jove Jove; fi
    sys.path.append('./Jove')
    sys.path.append('./Jove/jove')

# -- common imports --
from jove.PcpJupyterNew import *
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Own install


In [2]:
# Test and encure symlink is created successfully
os.path.isfile("./pcp.exe")

True

<hr class="double-div"/>

In [3]:
help(pcp_solve)

Help on function pcp_solve in module jove.PcpJupyterNew:

pcp_solve(pcp_pairs, OWN_INSTALL=False, run=None, ni=False, di=None, depth=None, tiles_per_row=15)
    Forward user input to a file, which we then use Ling Zhao's pcp solver to solve.
    :param OWN_INSTALL: if True, running on one's own machine; default is Colab (so false)
    :param pcp_pairs: List of tuple pairs representing pcp 'tiles'
    :param run: Number of runs to perform.
    :param ni: No iterative search.
    :param di: Depth increment.
    :param depth: Search depth.
    :param tiles_per_row: Number of tiles to show in single row together (useful for creating more meaningful output)



# (1) LT, 25%:
 

## 1(a): 

In [4]:
# TOol Solution
pcp_solve([('110', '1'), ('1', '0'), ('0', '110')], OWN_INSTALL)

 Running the command ... :  ['pcp.exe', '-i', 'temp.txt']
Solution(s) to PCP instance are below. Note: the tiles may be reversed,
  as the solver may sometimes present the solution in reverse.:

Solution 1
Length:  75
[3, 2, 2, 3, 3, 1, 3, 2, 2, 2, 1, 3, 3, 3, 2, 1, 3, 2, 2, 1, 3, 2, 1, 3, 3, 2, 1, 3, 2, 3, 2, 2, 1, 3, 2, 1, 3, 3, 2, 2, 3, 3, 2, 1, 3, 2, 1, 1, 3, 3, 1, 1, 3, 2, 1, 1, 1, 1, 2, 2, 1, 1, 2, 2, 3, 1, 3, 3, 1, 2, 1, 1, 2, 1, 1] 

0   1   1   0   0   011 0   1   1   1   011 0   0   0   1   
011 0   0   011 011 1   011 0   0   0   1   011 011 011 0   

011 0   1   1   011 0   1   011 0   0   1   011 0   1   0   
1   011 0   0   1   011 0   1   011 011 0   1   011 0   011 

1   1   011 0   1   011 0   0   1   1   0   0   1   011 0   
0   0   1   011 0   1   011 011 0   0   011 011 0   1   011 

1   011 011 0   0   011 011 0   1   011 011 011 011 1   1   
0   1   1   011 011 1   1   011 0   1   1   1   1   0   0   

011 011 1   1   0   011 0   0   011 1   011 011 1   011 011 
1

> $$ 75 $$

## 1(b): 

In [5]:
# Tool solution
pcp_solve([('110', '1'), ('1', '0'), ('0', '110'), ('0', '1')], OWN_INSTALL)

 Running the command ... :  ['pcp.exe', '-i', 'temp.txt']
Solution(s) to PCP instance are below. Note: the tiles may be reversed,
  as the solver may sometimes present the solution in reverse.:

Solution 1
Length:  5
[3, 2, 2, 4, 1] 

0   1   1   0   011 
011 0   0   1   1   




> $$ 5 $$

## 1(c, hand): 

> ### The Tiles:
> ```
>   [ 1 ,  2  ,  3  ]
>   | 1 | 101 |  0  |
>   | 0 |  1  | 110 |
> ```
> ### Solution By Hand:
> ```
>   [  2  , 1 ,  3  ]
>   | 101 | 1 |  0  |
>   |  1  | 0 | 110 |
> ```

## 1(c, tool check): 

In [6]:
# Solution by tool for the same
pcp_solve([('1', '0'), ('101', '1'), ('0', '110')], OWN_INSTALL)

 Running the command ... :  ['pcp.exe', '-i', 'temp.txt']
Solution(s) to PCP instance are below. Note: the tiles may be reversed,
  as the solver may sometimes present the solution in reverse.:

Solution 1
Length:  3
[3, 1, 2] 

0   1   101 
011 0   1   




> $$ 3 $$

## 1(d): 

In [7]:
# Create the tiles separately; easier to change and run that way
# Found it hard to create own instances that explode   

TILES =  [ ('001', '0'), ('0', '1'), ('1', '001') ]
pcp_solve( TILES  , OWN_INSTALL)  

 Running the command ... :  ['pcp.exe', '-i', 'temp.txt']
Solution(s) to PCP instance are below. Note: the tiles may be reversed,
  as the solver may sometimes present the solution in reverse.:

Solution 1
Length:  75
[3, 2, 2, 3, 3, 1, 3, 2, 2, 2, 1, 3, 3, 3, 2, 1, 3, 2, 2, 1, 3, 2, 1, 3, 3, 2, 1, 3, 2, 3, 2, 2, 1, 3, 2, 1, 3, 3, 2, 2, 3, 3, 2, 1, 3, 2, 1, 1, 3, 3, 1, 1, 3, 2, 1, 1, 1, 1, 2, 2, 1, 1, 2, 2, 3, 1, 3, 3, 1, 2, 1, 1, 2, 1, 1] 

1   0   0   1   1   100 1   0   0   0   100 1   1   1   0   
100 1   1   100 100 0   100 1   1   1   0   100 100 100 1   

100 1   0   0   100 1   0   100 1   1   0   100 1   0   1   
0   100 1   1   0   100 1   0   100 100 1   0   100 1   100 

0   0   100 1   0   100 1   1   0   0   1   1   0   100 1   
1   1   0   100 1   0   100 100 1   1   100 100 1   0   100 

0   100 100 1   1   100 100 1   0   100 100 100 100 0   0   
1   0   0   100 100 0   0   100 1   0   0   0   0   1   1   

100 100 0   0   1   100 1   1   100 0   100 100 0   100 100 
0

<hr class="double-div"/>

# (2a) AR, 15%:
 
 
> - Suppose there exists a decider $H$ for $Halt_{TM}$. 
>   - $H$ expects to be given a Turing Machine $M$ and a string $w$ (_i.e._ $\langle M,\, w \rangle$).
> - Build a Program $D$ as follows:
>   1. $D$ takes a single argument $M$.
>   0. As its first step $D$ invokes $A$ on $\langle M,\, M \rangle$.
>   0. If $H\left(\langle M,\, M \rangle\right)$ rejects $\longrightarrow$ $D\left(M\right)$ accepts.
>   0. If $H\left(\langle M,\, M \rangle\right)$ loops $\longrightarrow$ $D\left(M\right)$ rejects.
> - Now let's determine what $D\left(\langle D\rangle\right)$ results in:
>   - $D\left(\langle D\rangle\right)$ $\equiv$ $H\left(\langle D,\, D\rangle\right)$
>   - Suppose $H\left(\langle D,\, D\rangle\right)$ rejects $\longrightarrow$ 
>      $D\left(\langle D\rangle\right)$ accepts.
>     - However $H$ is a decider for $Halt_{TM}$, 
>        $\therefore$ the fact that $H\left(\langle D,\, D\rangle\right)$ rejects,
>        means that $D$ is a turring machine that fails-to-halt/loops on $\langle D\rangle$.
>     - _i.e._ $D\left(\langle D\rangle\right)$ rejects or loops
>     - Which is a contradiction.
>   - Suppose $H\left(\langle D,\, D\rangle\right)$ accepts $\longrightarrow$ 
>      $D\left(\langle D\rangle\right)$ rejects.
>     - However $H$ is a decider for $Halt_{TM}$, 
>        $\therefore$ the fact that $H\left(\langle D,\, D\rangle\right)$ rejects,
>        means that $D$ is a turring machine that halts on $\langle D\rangle$.
>     - _i.e._ $D\left(\langle D\rangle\right)$ accepts
>     - Which is also a contradiction.
> - Therefore, we obtain a contradiction under all cases. 
> - Thus, $Halt_{TM}$ is undecidable.
 

# (2b) AR, 10%:

> - The mapping reduction from $A_{TM}$ to $CFL_{TM}$ is to emit this machine $M'$ 
>     ```
>       M'(x) {
>           if x is of the form "0^n 1^n 0^n" then goto accept_M' ;
>           Run M on w ;
>           If M accepts w, goto accept_M' ;
>           If M rejects w, goto reject_M' ; }
>     ```
> 
> - Suppose $M'$ has a CFL.
> - Then given a decider for $CFL_{TM}$ ($Decider_{CFL_{TM}}$):
>   - $Decider_{CFL_{TM}}$ will accept $M'$.
>     - Therefore we know that:
>         $$
>           M'\in CFL_{TM}
>         $$
>
> - Therefore if $f(x)=M'$ is a true maping reduction between $A_{TM}$ and $CFL_{TM}$,
>    then:
>     $$ 
>       f(x)=M' \in CFL_{TM} \;\forall x\in L_x \\
>         % \{ x \::\: x \text{ is a string representation of } \langle M,\,w\rangle\in A_{TM} \} \\
>               \land \\
>       f(x)=M' \notin CFL_{TM} \;\forall x\in \overline{L_x}
>         % \{ x \::\: x \text{ is a string representation of } \langle M,\,w\rangle\notin A_{TM} \}
>     $$
>    Where $L_x$ is defined as:
>     $$ 
>       L_x = \{ x \::\: x \text{ is a string representation of } \langle M,\,w\rangle\in A_{TM} \}
>     $$
>    - $L_x$ is known to be regular of the form $\Sigma^*$
>
> - We can use $Decider_{CFL_{TM}}(M')$ to determine this for all $M'$.
>   - This would be exhasive to acctualy do so we just need to prove that it is the case.
>
> - To do this we will determine all the differnet clasification of languages that $M'$ created by $f$ 
>    can have $\forall x\in\Sigma^*$ can take, by examining the algorithm outlined above.
>
>   - Given any $x\in\Sigma^*$ representing $\langle M,\,w\rangle$,
>      $M'$ will only accept $x$ if $x\in\{0^n1^n0^n\::\:n\geq0\}$ or if $M$ accepts $w$
>      (_i.e._ $x\in L_x$). 
>     - This means that $M'$ will never accept any $x$ representing $\langle M,\,w\rangle$
>        where $M$ rejects or loops on $w$, unless it takes the form $0^n1^n0^n$.
>
>   - From this reasoning we get the following logic:
>     $$
>       Decider_{CFL_{TM}}(M')  = \left\{ \begin{array}{rcl} ~\\
>           \text{accepts}  & \Rightarrow & L(M') \text{ is Context-Free} \\
>                           & \Rightarrow & \text{Language is } \Sigma^* \\
>                           & \Rightarrow & L(M') \text{ is also Regular} \\
>                           & \Rightarrow & M \text{ accepts } w \\
>                       \\
>           \text{rejects}  & \Rightarrow & L(M') \text{ is not Context-Free} \\
>                           & \Rightarrow & \text{Language is } 0^n1^n0^n \\
>                           & \Rightarrow & L(M') \text{ is not Context-Sensitive} \\
>                           & \Rightarrow & M \text{ does }\textbf{not}\text{ accept } w
>       \end{array} \right.
>     $$
>
> - Thus we have achieved our mapping reduction 
>   - That is, $M'$ has a CFL if and only if $M$ accepts $w$.


<!-- $$
>       Decider_{CFL_{TM}}(M')  = \left\{ \begin{array}{rcl} ~\\
>           \text{accepts}    & \Rightarrow & L(M') \text{ is Context-Free} \\
>                             & \Rightarrow & \text{Language is } \Sigma^* \\
>                             & \Rightarrow & M \text{ accepts } w \\
>                       \\
>           \text{rejects}    & \Rightarrow & L(M') \text{ is not Context-Free} \\
>                             & \Rightarrow & \text{Language is } 0^n1^n0^n \\
>                             & \Rightarrow & M \text{ does not accept } w
>       \end{array} \right.
>     $$
-->

<!--
> - We can decide if a $f(x)$ is truly a maping reduction from $A_{TM}$ to $CFL_{TM}$.
>   - Which is true when for all $\langle M,\, w\rangle\in A_{TM}$ have an $M'$ gained thorough $f(x)$,
>      where $x$ is a string represenation of $\langle M,\, w\rangle$,
>      that are in $CFL_{TM}$, aka who's language they accept is a _CFL_.
> - In the defintion of $M'$ we can see that it's lanuage is defined such that:
>   - It will always accept strings of the form $0^n1^n0^n$ by themselves are a Contextless Language.
>   - Therefore $L(M')$ contain all strings $x$ resulting from $\langle M,\, w\rangle$ 
>      that take the form of $0^n1^n0^n$
>      (_i.e._ $x\in\{0^n1^n0^n \::\: n\geq0\}$)
>       $$
>         L(M') \:\supseteq\: \{0^n1^n0^n \::\: n\geq0\}
>       $$
>   - It will also accept if $M$ accepts $w$ reagardless of if $x$ is of the form $0^n1^n0^n$ or not.
>      (_i.e._ $M'$ accepts if $\langle M,\, w\rangle\in A_{TM}$).
>     - Therefore $L(M')$ contain all strings $x$ that represent 
>        $\langle M,\, w\rangle\in A_{TM}$, which is known to be "all else" _i.e._ $\Sigma^*$.
>         $$
>           L(M') \:\supseteq\: \Sigma^*
>         $$  
>     - This means that $L(M')$ is Regular, and all CFL's are regular languages.
>       - $\therefore$ $L(M')$ is also Context Free.
>   - It will reject if $M$ rejects $w$ unless its $x$ is of the form $0^n1^n0^n$
>      (_i.e._ $M'$ rejects if $\langle M,\, w\rangle\notin A_{TM}$)
>     - This means that $\langle M,\, w\rangle\notin A_{TM}$ then 
>        $L(M')$ only containes strings of the form $0^n1^n0^n$.
>         $$
>           L(M') \:\supseteq\: \{0^n1^n0^n \::\: n\geq0\}
>         $$  
>     - This means that $L(M')$ is contextual (CL), and not all CLs are CFLs.
>       - $\therefore$ $L(M')$ is **not** a CFL.
> - Therefore we know that any $x$ representing a $\langle M,\, w\rangle$ where $M$ rejects $w$,
>     will be rejected and not be part of $L(M')$.
  -->
 
 

<hr class="double-div"/>

# (3, XL, SV) 50%:  


## 3.a.i - Describe $\mathbf{A},\, \mathbf{B},\, f$:

- What is $\mathbf{A}$ in this problem? Describe clearly.
    > A language known to be undecidable.
    > $$ \mathbf{A} \subseteq \Sigma^* $$
    > <br/>
    > Which for this problem would be:
    > $$ 
    >   PCP = \left\{ 
    >     S 
    >       \;:\; 
    >     S \text{ is a finite sequience of elements over } \mathcal{T} \text{ that has a solution}
    >   \right\} 
    > $$
    > Where $\mathcal{T}$ is the _tile alphabet_ defined as:
    > $$
    >   \mathcal{T} \subseteq \Sigma^+ \times \Sigma^+  \\
    >     \land \\
    >   |\Sigma|>1
    > $$

- What is $\mathbf{B}$ in this problem? Describe clearly.
    > A second list of words over a finite alphabet $\Sigma$.
    > $$ \mathbf{B} \subseteq \Sigma^* $$
    > <br/>
    > Which for this problem would be:
    > $$ 
    >   Amb = \left\{ 
    >     \langle G\rangle 
    >       \;:\; 
    >     G \text{ is a } CFG \text{ that has } \geq1 \text{ parse for some } w\in\Sigma^* 
    >   \right\} 
    > $$

- What is $f$ in this problem? Describe clearly.
    > A _computable_ function, that serves as a maping reduction.
    > $$ f: \Sigma^* \rightarrow \Sigma^* $$
    > So that all strings ($x$) in $\mathbf{A}$ have a corosponding string $f(x)$ in $\mathbf{B}$,
    >  and all strings ($x$) not in $\mathbf{A}$ have no corospondign string $f(x)$ in $\mathbf{B}$.
    > $$ 
    >   \left( x\in\mathbf{A} \leftrightarrow f(x)\in\mathbf{B} \right)\: \forall x\in\Sigma^* 
    > $$
    > 
    > Which for this problem is function that generates _CFG_ $G$,
    >  based on a string $S$ that represents a set of tiles from the PCP problem,
    >  which is ambigiouse if and only if those tiles can produce a solution to the PCP problem.
    > $$
    >   \left( S\in PCP \:\leftrightarrow\: f(S)=G\in Amb \right)\: \forall S\in\mathcal{T}^*
    > $$
    >
    > In this case the framework for the _CFG_ $G$ that is constructed by $f$ 
    >  given a string $x$ that represents a finite sequince of elements over $\mathcal{T}$ ($S$):
    > $$
    >   f(x) = G = \left(\begin{array}{rl}  ~\\
    >       \mathbf{N}:         &   \{\mathcal{S},\, \mathcal{S}_A,\, \mathcal{S}_B\},    \\
    >       \mathbf{\Sigma}:    &   \Sigma\cup\{a_1,\, \dots,\, a_n\},  \\[2pt]
    >       \mathbf{P}:         &   \left\{\begin{array}{rcll}
    >                                   \mathcal{S}     & \rightarrow   
    >                                                   & \mathcal{S}_A \:|\: \mathcal{S}_B
    >                                                       &   \\
    >                                   \mathcal{S}_A   & \rightarrow
    >                                                   & w_i\mathcal{S}_Aa_i \:|\: w_ia_i
    >                                                       & \forall (i\leq n)\in\mathbb{N}  \\
    >                                   \mathcal{S}_B   & \rightarrow
    >                                                   & x_i\mathcal{S}_Ba_i \:|\: x_ia_i
    >                                                       & \forall (i\leq n)\in\mathbb{N}  \\
    >                               \end{array}\right\},\\[2pt]
    >       \mathbf{S}:         &   \mathcal{S}     \\[-8pt]~
    >   \end{array}\right)
    > $$
    > Where $A$ and $B$ are defined as in the question promt,
    >  meaning that $A=w_1,\,...,\,w_n$ is a list of strings $S\in PCP$ 
    >  that represent one side of the tiles, in the order they appear in and
    >  $B=x_1,\,...,\,x_n$ represents the same but for other side of the tiles.
    > Additionaly $a_1,\,...,\,a_n$ is a list of unique characters/symbols,
    >  that serve as id markers of sorts for identifying the order of the tiles.
    >
    > Finaly the _CFG_ $G$ will be run on a string $s$ that represents the concatination of 
    >  all of the strings in $A$ followed by all of the strings in $a_1,\,...,\,a_n$
    >  _(**Please Note:** this could also be done with the concatination of
    >   $B$ followed by all of the strings in $a_1,\,...,\,a_n$, and it would be functionaly the same)_.
    > This means that the _teminal alphabet_ of $G$ ($\mathbf{\Sigma}$) must be the union of $\Sigma$ 
    >  as defined in the defiition of the _tile alphabet_ ($\mathcal{T}$).
    >
    > All this means that $G$ will accept any string that represent any concatination of the
    >  strings that make up the top or bottom of the tiles in a PCP problem, 
    >  repeats allowed & order is irelivant,
    >  concatinated with a string of symbols from $a_1,\,...,\,a_n$,
    >  that represent the order of the tile strings concatinated together before, 
    >  _**but in reverse order**_.

<br/>

## 3.a.ii - Proof for the $x \in \mathbf{A}  \Rightarrow  f(x) \in \mathbf{B}$ part:

> - Suppose $S\in PCP$ (_aka_ $x\in\mathbf{A}$)
> - Then $f(S)=G$ such that $A$ and $B$ are defined as the top and bottom of the tiles
>    in a solution to the PCP problem.
>   - Meaning that the concatination of the strings in $A$ is equal the same as the
>      the concatination of the strings in $B$.
> - Therefore, as defined/explained in the section on $f$ above, 
>    $G$ will accept any string $s$ that represents a valid concatination of the strings in $A$,
>    followed by the concatination of the symbols from $a_1,\,...,\,a_n$ that correctly
>    match the order of tiles that are nessisary to generate the aformentioned string.
> - However if $G$ was generated from a valid solution $S\in PCP$ (_aka_ what we are assumeing),
>    then there exists by definition the definition of a solution to the PCP problem
>    a string that will have two possible parses to be accepted by $G$
>    (_aka_ $G$ will be ambiguiouse).
>   - We know this because the constant structure of $G$, 
>      no matter what is fed into $f$,
>      is defined such that it will try and parse it in terms of $A$ or $B$
>      (_aka_ in terms of the tops or bottoms of the tiles),
>   - But the solution to a PCP problem is such that the string resulting form both 
>      the tops of the tiles and the bottoms of the tiles will be the same when placed
>      in a certain order with repeats.
>   - Therefore, if $G$ is constructed from a set of tiles in $PCP$ 
>      (_aka_ sets of tiles that have a solution to the PCP problem, _i.e._ $S$),
>      then $G$ will by definition have a possible parse from both the $\mathcal{S}_A$
>      $\mathcal{S}_B$ route that will both be accepted for any string that 
>      represents a solution to the PCP problem with the tiles $S$ used to generate $G$.
>     - Since the solution string would represent the resulting string and the order of tiles
>        for both top and bottom, the parse atempt generated by the $\mathcal{S}_A$ or $\mathcal{S}_B$
>        possibility branches would not terminate, since both the top and the bottom conncatinate
>        in the same order to get the string represented in $s$.
> - Therefore, $G$ must be an ambigiouse grammar (inherent in the definition).
>     $$ G\in Amb \\ \therefore\; f(x)\in \mathbf{B} $$
> - This proves:
>     $$ S \in PCP  \Rightarrow  f(S)=G \in Amb $$
> - _aka_
>     $$ x \in \mathbf{A}  \Rightarrow  f(x) \in \mathbf{B} $$

 
<br/>

## 3.b - Proof for the $x \notin \mathbf{A} \Rightarrow f(x) \notin \mathbf{B}$ part:
 
> - Suppose $S\notin PCP$ (_aka_ $x\notin\mathbf{A}$)
> - Then $f(S)=G$ such that $A$ and $B$ are defined as the top and bottom of the tiles
>    in a solution to the PCP problem.
>   - Meaning that the concatination of the strings in $A$ is equal the same as the
>      the concatination of the strings in $B$.
> - Therefore, as defined/explained in the section on $f$ above, 
>    $G$ will accept any string $s$ that represents a valid concatination of the strings in $A$,
>    followed by the concatination of the symbols from $a_1,\,...,\,a_n$ that correctly
>    match the order of tiles that are nessisary to generate the aformentioned string.
> - As discussed/explained/prooved in the previouse answer we know that a possible way for $G$
>    to be generated from $f$ such that it is ambiguiouse is only possible when $G$ is generated from
>    an $S\in PCP$ and we are defining this part of the proof to exclude that as an option.
> - Therefore we just need to proove that there is no other way for $f$ to generate an ambigiouse $G$
>    other than what was defined previously to proove that $S \notin PCP\Rightarrow f(S)=G \notin Amb$.
>   - Due to the preceeding logic any possible ambiguity in $G$ will have to come from some 
>      product rule other than $\mathcal{S}\rightarrow\mathcal{S}_A\:|\:\mathcal{S}_B$.
>   - Therefore we must examine the product rules that are generated by $f$:
>     - Every other product rule that is constructed by $f$ is set up so that the parse must
>        start by using whatever product rule containes the $a_i$ that is the last char $s$,
>        in combination with the string $w_i$ or $x_i$ that must start $s$,
>        and cascade down parsing teh string from teh outsides in,
>        respecting if it is a $\mathcal{S}_A$ or $\mathcal{S}_B$ production rule, respectivly.
>     - This means that there can be no ambiguity since all $a_i$ are unique and tied to 
>        exactly one string $w_i$ and $x_i$, in all production rules being examined, 
>        and no grammars in any production rule in question that does not conform to this pattern.
> - Therefore, $G$ can **not** be an ambigiouse grammar (inherent in the structure).
>     $$ G\in Amb \\ \therefore\; f(x)\in \mathbf{B} $$
> - This proves:
>     $$ S \notin PCP  \Rightarrow  f(S)=G \notin Amb $$
> - _aka_
>     $$ x \notin \mathbf{A}  \Rightarrow  f(x) \notin \mathbf{B} $$


<hr class="double-div"/>

# END
